akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -702,6 +702,13 @@ sealed abstract class ByteString extends IndexedSeq[Byte] with IndexedSeqOptimiz
   // optimized in subclasses
   override def indexOf[B >: Byte](elem: B): Int = indexOf(elem, 0)
 
+  override def grouped(size: Int): Iterator[ByteString] = {
+    require(size >= 1, s"size=$size must be positive")
```

Comment:
This is a bit heavy -- it allocates a function for the string creation; we try to do the if manually in such perf sensitive pieces (well, one could argue grouped is by itself not as crazy sensitive in the call itself, however since the entire bytestring is optimized as well, let's optimise this too)

Reply:
Yeah change it please; scala watches out less for allocs than we kind of, so if we can improve let's do so. Such small things to build up to alloc rate in bigger systems after all

Reply:
> Some wise person once said "Trust no one, bench everything."

Very good :-)

<img src="https://user-images.githubusercontent.com/120979/40705492-fd411e0c-6425-11e8-9484-fbbffde6e7d6.png" width=300px/>

> JFR/JMC tracks zero allocations over a 60s period. Same thing over a longer 2m benchmark. The used heap line doesn't budge.
> 
> Is the jvm is doing something smart? Is my benchmark flawed? I dunno.

Hmm...yeah I wonder... Maybe using something else than a constant in the => String would help make sure it does not get optimized away and replaced with a constant?

I think we can replace it anyway since that's basically something that's pretty common in any perf touching code that we do, does not complicate the code too much etc.

Reply:
Same results for:
```scala
@Benchmark
def requireBench(bh: Blackhole, state: AkkaState) = {
  bh.consume(require(System.currentTimeMillis() > 0, s"${bh.consume(42)} $bh $state"))
}
```

![first world jvm problems](https://imgur.com/jWL8Alj.jpg)

Reply:
I see, we could keep digging but I think we're fine to just do it in this case to be honest :)

Reply:
Sure, fine by me.